src/common/stats/StatsManager.cpp

```diff
@@ -309,6 +321,214 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
+        std::ostream& out, const std::string& metricName, const std::string& metricType) const {
+    out << "# HELP " << metricName << " Record all " << metricType << " about nebula" << "\n";
+    out << "# TYPE " << metricName << " " << metricType << "\n";
+}
+
+// Write a double as a string, with proper formatting for infinity and NaN
+void PrometheusSerializer::writeValue(std::ostream& out, double value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        auto saved_flags = out.setf(std::ios::fixed, std::ios::floatfield);
+        out << value;
+        out.setf(saved_flags, std::ios::floatfield);
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const int64_t value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        out << value;
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const std::string& value) const {
+    for (auto c : value) {
+        switch (c) {
+        case '\n':
+            out << '\\' << 'n';
+            break;
+
+        case '\\':
+            out << '\\' << c;
+            break;
+
+        case '"':
+            out << '\\' << c;
+            break;
+
+        default:
+            out << c;
+            break;
+        }
+    }
+}
+
+template<typename T>
+void PrometheusSerializer::writeHead(std::ostream& out, const std::string& family,
+        const std::vector<Label>& labels, const std::string& suffix,
+        const std::string& extraLabelName,
+        const T extraLabelValue) const {
+    out << family << suffix;
+    if (!labels.empty() || !extraLabelName.empty()) {
+        out << "{";
+        const char* prefix = "";
+
+        for (auto& lp : labels) {
+            out << prefix << lp.name << "=\"";
+            writeValue(out, lp.value);
+            out << "\"";
+            prefix = ",";
+        }
+        if (!extraLabelName.empty()) {
+            out << prefix << extraLabelName << "=\"";
+            writeValue(out, extraLabelValue);
+            out << "\"";
+        }
+        out << "}";
+    }
+    out << " ";
+}
+
+// Write a line trailer: timestamp
+void PrometheusSerializer::writeTail(std::ostream& out, const std::int64_t timestampMs) const {
+    if (timestampMs != 0) {
+        out << " " << timestampMs;
+    }
+    out << "\n";
+}
+
+
+void PrometheusSerializer::prometheusSerialize(std::ostream& out, StatsManager& sm) const {
+    std::locale saved_locale = out.getloc();
+    out.imbue(std::locale::classic());
+    for (auto& index : sm.nameMap_) {
+        auto parsedName = sm.idParsedName_[index.second].get();
+        std::string name = parsedName != nullptr ? parsedName->name : index.first;
+        if (StatsManager::isStatIndex(index.second)) {
+            annotate(out, name, "gauge");
```

Comment:
Here should be counter, Gauge represent instantaneous value.

Reply:
reference: https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#basic-info

Reply:
But this will decrease.

Reply:
OK, qps actually could decrease in some times.